fastapi_oidc/auth.py

```diff
@@ -1,125 +1,266 @@
-            server.
-        base_authorization_server_uri(URL): Everything before /.wellknow in your auth
-            server URL. I.E. https://dev-123456.okta.com
-        issuer (URL): Same as base_authorization. This is used to generating OpenAPI3.0
-            docs which is broken (in OpenAPI/FastAPI) right now.
-        signature_cache_ttl (int): How many seconds your app should cache the
-            authorization server's public signatures.
-        audience (str): (Optional) The audience string configured by your auth server.
-            If not set defaults to client_id
-        token_type (IDToken or subclass): (Optional) An optional class to be returned by
-            the authenticate_user function.
-
-
-    Returns:
-        func: authenticate_user(auth_header: str) -> IDToken (or token_type)
-
-    Raises:
-        Nothing intentional
-    """
-
-    if not issubclass(token_type, IDToken):
-        raise TokenSpecificationError(
-            "Invalid argument for token_type. "
-            "Token type must be a subclass of fastapi_oidc.type.IDToken. "
-            f"Received {token_type=}"
+from fastapi_oidc.grant_types import GrantType
+from fastapi_oidc.idtoken_types import IDToken
+
+
+class Auth(OAuth2):
+    def __init__(
+        self,
+        openid_connect_url: str,
+        issuer: Optional[str] = None,
+        client_id: Optional[str] = None,
+        scopes: List[str] = list(),
+        grant_types: List[GrantType] = [GrantType.IMPLICIT],
+        signature_cache_ttl: int = 3600,
+        idtoken_model: Type = IDToken,
+    ):
+        """Configure authentication and use method :func:`require` or :func:`optional`
+        to check user credentials.
+
+        Args:
+            openid_connect_url (URL): URL to the "well known" openid connect config
+                e.g. https://dev-123456.okta.com/.well-known/openid-configuration
+            issuer (URL): (Optional) The issuer URL from your auth server.
+            client_id (str): (Optional) The client_id configured by your auth server.
+            scopes (Dict[str, str]): (Optional) A dictionary of scopes and their descriptions.
+            grant_types (List[GrantType]): (Optional) Grant types shown in docs.
+            signature_cache_ttl (int): (Optional) How many seconds your app should
+                cache the authorization server's public signatures.
+            idtoken_model (Type): (Optional) The model to use for validating the ID Token.
+
+        Raises:
+            Nothing intentional
+        """
+
+        self.openid_connect_url = openid_connect_url
+        self.issuer = issuer
+        self.client_id = client_id
+        self.idtoken_model = idtoken_model
+        self.scopes = scopes
+
+        self.discover = discovery.configure(cache_ttl=signature_cache_ttl)
+        oidc_discoveries = self.discover.auth_server(
+            openid_connect_url=self.openid_connect_url
+        )
+        # scopes_dict = {
```

Comment:
The flows take this argument but I don't think it changes anything. I left the code there because I'm not sure and because it would be nice to let the user decide what scopes to show. At the moment we are discovering the possible scopes and it can be a lot

Reply:
If we're not going to use it right now I'd advocate we remove it. I'm open to adding an Issue link next to it and leaving it in if you have plans to implement or investigate.

Reply:
Sure let's remove for now